csrc/abstract_tensor.h

```diff
@@ -332,10 +849,139 @@ struct AbstractTensor {
+    }
+    return *this;
+  }
 
   // Remove all the null elements.
-  AbstractTensor& strip();
+  AbstractTensorWithInfo& strip() {
+    AbstractTensorWithInfo result;
+    for (const auto& aid : domain) {
+      if (aid.hasValue()) {
+        result.pushBack(aid);
+      }
+    }
+    std::swap(result, *this);
+    return *this;
+  }
+};
+
+struct EmptyInfo {
+  static EmptyInfo merge(const EmptyInfo&, const EmptyInfo&) {
+    return {};
+  }
+
+  static std::pair<EmptyInfo, EmptyInfo> split(const EmptyInfo& a) {
+    return {{}, {}};
+  }
+
+  static std::pair<EmptyInfo, EmptyInfo> swizzle(
+      const EmptyInfo& a,
+      const EmptyInfo& b) {
+    return {{}, {}};
+  }
+
+  bool operator==(const EmptyInfo& t) const {
+    return true;
+  }
+};
+
+using AbstractTensor = AbstractTensorWithInfo<EmptyInfo>;
+
+//! This is a holds a comparable type of "tag". This version is strict in the
+//! sense that we check that the tags match whenever we do a merge instead of
+//! allowing mismatches.
+template <typename Tag>
+struct TagSetInfo {
+  std::unordered_set<Tag> tags;
+
+  TagSetInfo(std::unordered_set<Tag> tags_) : tags(tags_) {}
+
+  static TagSetInfo<Tag> merge(
+      const TagSetInfo<Tag>& a,
+      const TagSetInfo<Tag>& b) {
+    TagSetInfo<Tag> merged_tag_info{a.tags};
+    merged_tag_info.tags.insert(b.tags.begin(), b.tags.end());
+    return merged_tag_info;
+  }
+
+  static std::pair<TagSetInfo<Tag>, TagSetInfo<Tag>> split(
+      const TagSetInfo<Tag>& a) {
+    return {a, a};
+  }
+
+  //! Swizzling mixes the tags so here we re-use merge and duplicate the result
+  static std::pair<TagSetInfo<Tag>, TagSetInfo<Tag>> swizzle(
+      const TagSetInfo<Tag>& a,
+      const TagSetInfo<Tag>& b) {
+    TagSetInfo<Tag> merged_info = merge(a, b);
+    return {merged_info, merged_info};
+  }
+
+  bool operator==(TagSetInfo<Tag>&& t) const {
+    return tags == t.tags;
+  }
+};
+
+//! This is a special case of AbstractTensorWithInfo which propagates a set of
+//! tags for each axis. The tags can be any hashable type with equality;
+//! typically Tag would be an enum.
+template <typename Tag>
+struct TaggedAbstractTensor : AbstractTensorWithInfo<TagSetInfo<Tag>> {
+  TaggedAbstractTensor(
+      std::vector<AbstractId> domain,
+      const std::vector<std::unordered_set<Tag>>& tag_sets)
+      : AbstractTensorWithInfo<TagSetInfo<Tag>>(
+            domain,
+            {tag_sets.begin(), tag_sets.end()}) {}
+  TaggedAbstractTensor(
+      const std::vector<IterDomain*>& domain,
+      const std::vector<std::unordered_set<Tag>>& tag_sets)
+      : AbstractTensorWithInfo<TagSetInfo<Tag>>(
+            domain,
+            {tag_sets.begin(), tag_sets.end()}) {}
+  TaggedAbstractTensor(
+      std::initializer_list<AbstractId> domain,
+      std::initializer_list<std::initializer_list<Tag>> tag_sets)
+      : AbstractTensorWithInfo<TagSetInfo<Tag>>(
+            domain,
+            {tag_sets.begin(), tag_sets.end()}) {}
```

Comment:
This actually didn't work for me because although there is an `unordered_set` ctor for `TagSetInfo`, there's no `vector<unordered_set<Tag>>` ctor for `vector<TagSetInfo>`.